### PR TITLE
Remove unwrap_none/expect_none from compiler/.

### PR DESCRIPTION
**Pull request by m-ou-se**
We're not going to stabilize `Option::{unwrap_none, expect_none}`. (See https://github.com/rust-lang/rust/issues/62633.) This removes the usage of those unstable methods from `compiler/`.

### CODE REVIEW

**Comment by petrochenkov** [comment]
Could you avoid putting anything with side effects into asserts?
Asserts may be casually removed or turned into debug asserts, so running meaningful code in them always triggers some level of alarm.

**Comment by m-ou-se** [comment]
Done.

I added `try_insert` (https://github.com/rust-lang/rust/pull/82764) and `assert_matches` (https://github.com/rust-lang/rust/pull/82770) to make it a bit nicer.

File: compiler/rustc_middle/src/ich/impls_syntax.rs [hunk 1]
```diff
@@ -45,7 +45,11 @@ impl<'ctx> rustc_ast::HashStableContext for StableHashingContext<'ctx> {
             item.hash_stable(self, hasher);
             style.hash_stable(self, hasher);
             span.hash_stable(self, hasher);
-            tokens.as_ref().expect_none("Tokens should have been removed during lowering!");
+            assert_matches!(
+                tokens.as_ref(),
+                None,
```
Comment (petrochenkov) [inline compiler/rustc_middle/src/ich/impls_syntax.rs:50]: FWIW, in all these cases with `None` `assert_eq` would work too.

**Comment by petrochenkov** [comment]
Could you avoid evaluating things twice as well?
```rust
foo(map[key]);
assert_eq!(map[key], expected);
```
=>
```rust
let value = map[key];
foo(value);
assert_eq!(value, expected);
```
It's not immediately obvious whether it's something cheap like array access or more expensive like `BTreeMap` search in each specific case.

**Comment by m-ou-se** [comment]
> Could you avoid evaluating things twice as well?

Done.

**Comment by petrochenkov** [comment]
r=me with commits squashed.

**Pull request merged**